rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/util/Diagnostics.java

```diff
@@ -91,71 +89,120 @@ public static interface Template {
+                skippedLoc.getOffset() - completeErrorTreeLoc.getOffset(),
+                completeErrorTreeLoc.getBeginLine(),
+                skippedLoc.getBeginLine(),
+                completeErrorTreeLoc.getBeginColumn(),
+                skippedLoc.getBeginColumn()
+        );
+
+        String nonterminal = SymbolAdapter.toString(TreeAdapter.getType(errorTree), false);
 
-        if (ERROR_TREE_HIGHLIGHT != null) {
-            // The whole error tree
-            diagnostics.add(cm -> new Diagnostic(toRange(errorTreeLoc, cm), "Recovered parse error", ERROR_TREE_HIGHLIGHT, "parser"));
-        }
+        List<Template> diagnostics = new ArrayList<>();
 
-        if (PREFIX_HIGHLIGHT != null) {
-            // The recognized prefix
-            int prefixLength = skippedLoc.getOffset()-errorTreeLoc.getOffset();
-            if (prefixLength > 0) {
-                ISourceLocation prefixLoc = factory.sourceLocation(errorTreeLoc,
-                        errorTreeLoc.getOffset(), skippedLoc.getOffset()-errorTreeLoc.getOffset(),
-                        errorTreeLoc.getBeginLine(), skippedLoc.getBeginLine(),
-                        errorTreeLoc.getBeginColumn(), skippedLoc.getBeginColumn());
-                diagnostics.add(cm -> new Diagnostic(toRange(prefixLoc, cm), "Recovered parse error prefix", PREFIX_HIGHLIGHT, "parser"));
+        diagnostics.add(cm -> {
+            var d = new Diagnostic(
+                toRange(stuckLoc, cm),
+                "Parsing got stuck here",
+                DiagnosticSeverity.Error,
+                "parser");
+
+            List<DiagnosticRelatedInformation> related = new ArrayList<>();
+
+            related.add(related(cm, stuckLoc,
+               "It is likely something is extra or missing here, or before this position. However, parsing can also have gone back here searching for a solution."));
+            related.add(related(cm, skippedLoc, "Parsing skipped this part to artificially complete a `" + nonterminal + "` and continue parsing. Something is missing or extra before the end of this part."));
+            related.add(related(cm, prefixLoc, "This part of the `" + nonterminal + "` was still recognized until parsing got stuck."));
+
+            // TODO: replace by ProductionAdapter calls once they are available with the new RC of Rascal
+            IConstructor errorProd = TreeAdapter.getProduction(errorTree);
+            int dot = ((IInteger) errorProd.get("dot")).intValue();
+            IConstructor prod = (IConstructor) errorProd.get("prod");
+
+            if (ProductionAdapter.isDefault(prod)) {
+                IList symbols = ProductionAdapter.getSymbols(prod);
+                IList prefix = symbols.sublist(0, dot);
+                IList postfix = symbols.sublist(dot, symbols.length());
+
+                related.add(related(cm, stuckLoc, "Experts: a `" + nonterminal + "` was the last but probably not the only grammar rule that was tried at this position."));
```

Comment:
can we move this experts information behind a settings flag? since it's a lot of information and might overwelm the user?

Reply:
Decided to remove it again. The incompleteness of it does not help an expert either.